diff --git a/options/internal/include/mlibc/thread-types.hpp b/options/internal/include/mlibc/thread-types.hpp
--- a/options/internal/include/mlibc/thread-types.hpp
+++ b/options/internal/include/mlibc/thread-types.hpp
@@ -12,9 +12,10 @@
 #include <new>
 #include <threads.h>
 #include <time.h>
+#include <mlibc/arch-defs.hpp>
 
 static constexpr size_t default_stacksize = 0x200000;
-static constexpr size_t default_guardsize = 4096;
+static constexpr size_t default_guardsize = mlibc::page_size;
 
 struct __mlibc_threadattr {
 	template <typename T>
