diff --git a/options/posix/generic/mqueue.cpp b/options/posix/generic/mqueue.cpp
--- a/options/posix/generic/mqueue.cpp
+++ b/options/posix/generic/mqueue.cpp
@@ -56,6 +56,14 @@ mqd_t mq_open(const char *name, int flags, ...) {
 	return out;
 }
 
+int mq_close(mqd_t mqdes) {
+	if(int e = mlibc::sysdep<Close>(mqdes); e) {
+		errno = e;
+		return -1;
+	}
+	return 0;
+}
+
 ssize_t mq_receive(mqd_t mqdes, char *msg_ptr, size_t msg_len, unsigned int *msg_prio) {
 	if (int e = mlibc::sysdep_or_enosys<MqReceive>(mqdes, msg_ptr, msg_len, msg_prio); e) {
 		errno = e;
diff --git a/options/posix/include/mqueue.h b/options/posix/include/mqueue.h
--- a/options/posix/include/mqueue.h
+++ b/options/posix/include/mqueue.h
@@ -18,2 +18,3 @@
 int mq_unlink(const char *__name);
+int mq_close(mqd_t __mqdes);
 mqd_t mq_open(const char *__name, int __flags, ...);
