diff --git a/options/rtld/generic/main.cpp b/options/rtld/generic/main.cpp
index 243de58b..c345589d 100644
--- a/options/rtld/generic/main.cpp
+++ b/options/rtld/generic/main.cpp
@@ -75,13 +75,12 @@ DebugInterface globalDebugInterface;
 // Use a PC-relative instruction sequence to find our runtime load address.
 uintptr_t getLdsoBase() {
-#if defined(__x86_64__) || defined(__i386__) || defined(__aarch64__) || defined(__m68k__) || defined(__loongarch64)
+#if defined(__aarch64__) || defined(__riscv)
+	return reinterpret_cast<uintptr_t>(&__ehdr_start);
+#elif defined(__x86_64__) || defined(__i386__) || defined(__m68k__) || defined(__loongarch64)
 	// On x86_64, the first GOT entry holds the link-time address of _DYNAMIC.
-	// TODO: This isn't guaranteed on AArch64, so this might fail with some linkers.
 	auto linktime_dynamic = reinterpret_cast<uintptr_t>(_GLOBAL_OFFSET_TABLE_[0]);
 	auto runtime_dynamic = reinterpret_cast<uintptr_t>(_DYNAMIC);
 	return runtime_dynamic - linktime_dynamic;
-#elif defined(__riscv)
-	return reinterpret_cast<uintptr_t>(&__ehdr_start);
 #else
 	#error Unknown architecture!
 #endif
