diff --git a/options/internal/include/mlibc/sysdep-signatures.hpp b/options/internal/include/mlibc/sysdep-signatures.hpp
--- a/options/internal/include/mlibc/sysdep-signatures.hpp
+++ b/options/internal/include/mlibc/sysdep-signatures.hpp
@@ -199,6 +199,8 @@ SYSDEP_FUNC(Sockname, int fd, struct sockaddr *addr_ptr, socklen_t max_addr_leng
 SYSDEP_FUNC(Peername, int fd, struct sockaddr *addr_ptr, socklen_t max_addr_length, socklen_t *actual_length);
 SYSDEP_FUNC(GetHostname, char *buffer, size_t bufsize);
 SYSDEP_FUNC(SetHostname, const char *buffer, size_t bufsize);
+SYSDEP_FUNC(GetDomainname, char *buffer, size_t bufsize);
+SYSDEP_FUNC(SetDomainname, const char *buffer, size_t bufsize);
 SYSDEP_FUNC(Mkfifoat, int dirfd, const char *path, mode_t mode);
 SYSDEP_FUNC(GetEntropy, void *buffer, size_t length);
 SYSDEP_FUNC(Mknodat, int dirfd, const char *path, int mode, int dev);
diff --git a/options/internal/include/mlibc/sysdep-tags.hpp b/options/internal/include/mlibc/sysdep-tags.hpp
--- a/options/internal/include/mlibc/sysdep-tags.hpp
+++ b/options/internal/include/mlibc/sysdep-tags.hpp
@@ -405,6 +405,10 @@ struct Peername {};
 // int sys_gethostname(char *buffer, size_t bufsize);
 struct GetHostname {};
 // int sys_sethostname(const char *buffer, size_t bufsize);
 struct SetHostname {};
+// int sys_getdomainname(char *buffer, size_t bufsize);
+struct GetDomainname {};
+// int sys_setdomainname(const char *buffer, size_t bufsize);
+struct SetDomainname {};
 // int sys_mkfifoat(int dirfd, const char *path, mode_t mode);
 struct Mkfifoat {};
 // int sys_getentropy(void *buffer, size_t length);
diff --git a/options/posix/generic/unistd.cpp b/options/posix/generic/unistd.cpp
--- a/options/posix/generic/unistd.cpp
+++ b/options/posix/generic/unistd.cpp
@@ -1548,4 +1548,7 @@
-int getdomainname(char *, size_t) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int getdomainname(char *buffer, size_t bufsize) {
+	if (int e = mlibc::sysdep<GetDomainname>(buffer, bufsize); e) {
+		errno = e;
+		return -1;
+	}
+	return 0;
 }
@@ -1553,4 +1556,7 @@
-int setdomainname(const char *, size_t) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int setdomainname(const char *buffer, size_t bufsize) {
+	if (int e = mlibc::sysdep<SetDomainname>(buffer, bufsize); e) {
+		errno = e;
+		return -1;
+	}
+	return 0;
 }
diff --git a/sysdeps/linux/generic/sysdeps.cpp b/sysdeps/linux/generic/sysdeps.cpp
--- a/sysdeps/linux/generic/sysdeps.cpp
+++ b/sysdeps/linux/generic/sysdeps.cpp
@@ -1417,9 +1417,29 @@ int Sysdeps<SetHostname>::operator()(const char *buffer, size_t bufsize) {
 	auto ret = do_syscall(SYS_sethostname, buffer, bufsize);
 	if (int e = sc_error(ret); e)
 		return e;
 	return 0;
 }
 
+int Sysdeps<GetDomainname>::operator()(char *buffer, size_t bufsize) {
+	struct utsname name;
+	auto ret = do_syscall(SYS_uname, &name);
+	if (int e = sc_error(ret); e)
+		return e;
+	size_t length = frg::min(bufsize, sizeof(name.domainname));
+	for (size_t i = 0; i < length; i++)
+		buffer[i] = name.domainname[i];
+	if (length > 0)
+		buffer[length - 1] = '\0';
+	return 0;
+}
+
+int Sysdeps<SetDomainname>::operator()(const char *buffer, size_t bufsize) {
+	auto ret = do_syscall(SYS_setdomainname, buffer, bufsize);
+	if (int e = sc_error(ret); e)
+		return e;
+	return 0;
+}
+
 int Sysdeps<Mkfifoat>::operator()(int dirfd, const char *path, mode_t mode) {
 	return sysdep<Mknodat>(dirfd, path, mode | S_IFIFO, 0);
 }
diff --git a/sysdeps/linux/include/mlibc/sysdeps.hpp b/sysdeps/linux/include/mlibc/sysdeps.hpp
--- a/sysdeps/linux/include/mlibc/sysdeps.hpp
+++ b/sysdeps/linux/include/mlibc/sysdeps.hpp
@@ -112,6 +112,8 @@ struct LinuxSysdepTags :
 	Sockname,
 	Peername,
 	GetHostname,
 	SetHostname,
+	GetDomainname,
+	SetDomainname,
 	Mkfifoat,
 	GetEntropy,
 	Mknodat,
