diff --git i/options/ansi/include/limits.h w/options/ansi/include/limits.h
index 40d36c71..9e8b9de0 100644
--- i/options/ansi/include/limits.h
+++ w/options/ansi/include/limits.h
@@ -71,6 +71,23 @@
 #define LINE_MAX 4096
 #define PIPE_BUF 4096
 
+#define CHAR_WIDTH __CHAR_BIT__
+#define SCHAR_WIDTH __CHAR_BIT__
+#define UCHAR_WIDTH __CHAR_BIT__
+#define SHRT_WIDTH __SHRT_WIDTH__
+#define USHRT_WIDTH __SHRT_WIDTH__
+#define INT_WIDTH __INT_WIDTH__
+#define UINT_WIDTH __INT_WIDTH__
+#define LONG_WIDTH __LONG_WIDTH__
+#define ULONG_WIDTH __LONG_WIDTH__
+#define LLONG_WIDTH __LLONG_WIDTH__
+#define ULLONG_WIDTH __LLONG_WIDTH__
+#define SIZE_WIDTH __SIZE_WIDTH__
+#define PTRDIFF_WIDTH __PTRDIFF_WIDTH__
+#define SIG_ATOMIC_WIDTH __SIG_ATOMIC_WIDTH__
+#define WCHAR_WIDTH __WCHAR_WIDTH__
+#define WINT_WIDTH __WINT_WIDTH__
+
 #define FILESIZEBITS 64
 
 #define CHARCLASS_NAME_MAX 14
