diff --git a/options/internal/include/mlibc/sysdep-signatures.hpp b/options/internal/include/mlibc/sysdep-signatures.hpp
--- a/options/internal/include/mlibc/sysdep-signatures.hpp
+++ b/options/internal/include/mlibc/sysdep-signatures.hpp
@@ -156,6 +156,7 @@ SYSDEP_FUNC(Linkat, int olddirfd, const char *old_path, int newdirfd, const char
 SYSDEP_FUNC(Symlink, const char *target_path, const char *link_path);
 SYSDEP_FUNC(Symlinkat, const char *target_path, int dirfd, const char *link_path);
 SYSDEP_FUNC(Renameat, int olddirfd, const char *old_path, int newdirfd, const char *new_path);
+SYSDEP_FUNC(Renameat2, int olddirfd, const char *old_path, int newdirfd, const char *new_path, unsigned int flags);
 SYSDEP_FUNC(Fcntl, int fd, int request, va_list args, int *result);
 SYSDEP_FUNC(Ttyname, int fd, char *buf, size_t size);
 SYSDEP_FUNC(Fadvise, int fd, off_t offset, off_t length, int advice);
diff --git a/options/internal/include/mlibc/sysdep-tags.hpp b/options/internal/include/mlibc/sysdep-tags.hpp
--- a/options/internal/include/mlibc/sysdep-tags.hpp
+++ b/options/internal/include/mlibc/sysdep-tags.hpp
@@ -316,5 +316,8 @@ struct Symlinkat {};
 // int sys_renameat(int olddirfd, const char *old_path, int newdirfd, const char *new_path);
 struct Renameat {};
+// int sys_renameat2(int olddirfd, const char *old_path, int newdirfd, const char *new_path,
+//     unsigned int flags);
+struct Renameat2 {};
 // int sys_fcntl(int fd, int request, va_list args, int *result);
 struct Fcntl {};
 // int sys_ttyname(int fd, char *buf, size_t size);
diff --git a/options/posix/generic/posix_stdio.cpp b/options/posix/generic/posix_stdio.cpp
--- a/options/posix/generic/posix_stdio.cpp
+++ b/options/posix/generic/posix_stdio.cpp
@@ -233,7 +233,10 @@ int renameat(int olddirfd, const char *old_path, int newdirfd, const char *new_pa
 	return 0;
 }
 
-int renameat2(int, const char *, int, const char *, unsigned int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int renameat2(int olddirfd, const char *old_path, int newdirfd, const char *new_path, unsigned int flags) {
+	if(int e = mlibc::sysdep_or_enosys<Renameat2>(olddirfd, old_path, newdirfd, new_path, flags); e) {
+		errno = e;
+		return -1;
+	}
+	return 0;
 }
diff --git a/sysdeps/linux/generic/sysdeps.cpp b/sysdeps/linux/generic/sysdeps.cpp
--- a/sysdeps/linux/generic/sysdeps.cpp
+++ b/sysdeps/linux/generic/sysdeps.cpp
@@ -1060,6 +1060,18 @@ int Sysdeps<Renameat>::operator()(int old_dirfd, const char *old_path, int new_di
 	return 0;
 }
 
+int Sysdeps<Renameat2>::operator()(int old_dirfd, const char *old_path, int new_dirfd,
+		const char *new_path, unsigned int flags) {
+#ifdef SYS_renameat2
+	auto ret = do_syscall(SYS_renameat2, old_dirfd, old_path, new_dirfd, new_path, flags);
+	if(int e = sc_error(ret); e)
+		return e;
+	return 0;
+#else
+	return ENOSYS;
+#endif
+}
+
 int Sysdeps<Fcntl>::operator()(int fd, int cmd, va_list args, int *result) {
 	auto arg = va_arg(args, unsigned long);
 	// TODO: the api for linux differs for each command so fcntl()s might fail with -EINVAL
diff --git a/sysdeps/linux/include/mlibc/sysdeps.hpp b/sysdeps/linux/include/mlibc/sysdeps.hpp
--- a/sysdeps/linux/include/mlibc/sysdeps.hpp
+++ b/sysdeps/linux/include/mlibc/sysdeps.hpp
@@ -70,6 +70,7 @@ enum class [[clang::enum_extensibility(open)]] Sysdep {
 	Symlink,
 	Symlinkat,
 	Renameat,
+	Renameat2,
 	Fcntl,
 	Ttyname,
 	Fadvise,
