diff --git i/options/ansi/include/math.h w/options/ansi/include/math.h
index 57a7995a..ae9fe66d 100644
--- i/options/ansi/include/math.h
+++ w/options/ansi/include/math.h
@@ -271,6 +271,27 @@ double lgamma(double __x);
 float lgammaf(float __x);
 long double lgammal(long double __x);
 
+#if defined(_DEFAULT_SOURCE) || __MLIBC_XOPEN || defined(_GNU_SOURCE)
+double j0(double __x);
+double j1(double __x);
+double jn(int __n, double __x);
+double y0(double __x);
+double y1(double __x);
+double yn(int __n, double __x);
+float j0f(float __x);
+float j1f(float __x);
+float jnf(int __n, float __x);
+float y0f(float __x);
+float y1f(float __x);
+float ynf(int __n, float __x);
+#endif
+
+#if defined(_DEFAULT_SOURCE) || defined(_GNU_SOURCE)
+double lgamma_r(double __x, int *__sign);
+float lgammaf_r(float __x, int *__sign);
+long double lgammal_r(long double __x, int *__sign);
+#endif
+
 double tgamma(double __x);
 float tgammaf(float __x);
 long double tgammal(long double __x);
@@ -327,6 +348,13 @@ double remainder(double __x, double __y);
 float remainderf(float __x, float __y);
 long double remainderl(long double __x, long double __y);
 
+#if defined(_DEFAULT_SOURCE) || defined(_GNU_SOURCE)
+double drem(double __x, double __y);
+float dremf(float __x, float __y);
+double scalb(double __x, double __power);
+float scalbf(float __x, float __power);
+#endif
+
 double remquo(double __x, double __y, int *__quotient);
 float remquof(float __x, float __y, int *__quotient);
 long double remquol(long double __x, long double __y, int *__quotient);
@@ -363,6 +391,11 @@ double fmin(double __x, double __y);
 float fminf(float __x, float __y);
 long double fminl(long double __x, long double __y);
 
+#if defined(_DEFAULT_SOURCE) || defined(_GNU_SOURCE)
+double significand(double __x);
+float significandf(float __x);
+#endif
+
 /* [C11/7.12.13 Floating multiply-add] */
 
 double fma(double __x, double __y, double __z);
