diff --git i/options/posix/generic/unistd.cpp w/options/posix/generic/unistd.cpp
index e013b40c..8cbe81c2 100644
--- i/options/posix/generic/unistd.cpp
+++ w/options/posix/generic/unistd.cpp
@@ -263,6 +263,8 @@ long fpathconf(int, int name) {
 	switch (name) {
 	case _PC_NAME_MAX:
 		return NAME_MAX;
+	case _PC_PATH_MAX:
+		return PATH_MAX;
 	case _PC_FILESIZEBITS:
 		return FILESIZEBITS;
 	default:
@@ -613,6 +615,8 @@ long pathconf(const char *, int name) {
 	switch (name) {
 	case _PC_NAME_MAX:
 		return NAME_MAX;
+	case _PC_PATH_MAX:
+		return PATH_MAX;
 	case _PC_FILESIZEBITS:
 		return FILESIZEBITS;
 	default:
