diff --git a/options/ansi/generic/stdlib.cpp b/options/ansi/generic/stdlib.cpp
--- a/options/ansi/generic/stdlib.cpp
+++ b/options/ansi/generic/stdlib.cpp
@@ -18,6 +18,7 @@
 #include <mlibc/all-sysdeps.hpp>
 #include <mlibc/allocator.hpp>
 #include <mlibc/charcode.hpp>
+#include <mlibc/locale.hpp>
 #include <mlibc/strtofp.hpp>
 #include <mlibc/strtol.hpp>
 #include <mlibc/threads.hpp>
@@ -32,6 +33,10 @@
 extern "C" int __cxa_atexit(void (*function)(void *), void *argument, void *dso_tag);
 void __mlibc_do_finalize();
 
+size_t __mlibc_mb_cur_max(void) {
+	return mlibc::getActiveLocale()->ctype.get(_NL_CTYPE_MB_CUR_MAX).asUint32();
+}
+
 namespace {
 	// According to the first paragraph of [C11 7.22.7],
 	// mblen(), mbtowc() and wctomb() have an internal state.
diff --git a/options/ansi/include/stdlib.h b/options/ansi/include/stdlib.h
--- a/options/ansi/include/stdlib.h
+++ b/options/ansi/include/stdlib.h
@@ -35,8 +35,8 @@ typedef struct {
 
 #define RAND_MAX 0x7FFFFFFF
 
-/* TODO: this should not be a compile-time constant */
-#define MB_CUR_MAX ((size_t)4)
+size_t __mlibc_mb_cur_max(void);
+#define MB_CUR_MAX (__mlibc_mb_cur_max())
 
 #ifndef __MLIBC_ABI_ONLY
 
