diff --git i/options/posix/generic/sys-wait.cpp w/options/posix/generic/sys-wait.cpp
index 092790a0..60bd2677 100644
--- i/options/posix/generic/sys-wait.cpp
+++ w/options/posix/generic/sys-wait.cpp
@@ -32,10 +32,12 @@ pid_t wait(int *status) {
 }
 
 pid_t wait3(int *status, int options, struct rusage *rusage) {
-	(void) rusage;
-	mlibc::infoLogger() << "\e[31mmlibc: wait3() is not implemented correctly\e[39m"
-		<< frg::endlog;
-	return waitpid(-1, status, options);
+	pid_t ret;
+	if(int e = mlibc::sysdep_or_enosys<Waitpid>(-1, status, options, rusage, &ret); e) {
+		errno = e;
+		return -1;
+	}
+	return ret;
 }
 
 pid_t wait4(pid_t pid, int *status, int options, struct rusage *ru) {
