diff --git a/options/posix/generic/dirent.cpp b/options/posix/generic/dirent.cpp
index 5f1ff23a..ab84a849 100644
--- a/options/posix/generic/dirent.cpp
+++ b/options/posix/generic/dirent.cpp
@@ -73,8 +73,8 @@ struct dirent *readdir(DIR *dir) {
 	__ensure(dir->__ent_next <= dir->__ent_limit);
 	if(dir->__ent_next == dir->__ent_limit) {
 		if(int e = mlibc::sysdep_or_enosys<ReadEntries>(dir->__handle, dir->__ent_buffer, 2048, &dir->__ent_limit); e)
-			__ensure(!"mlibc::sys_read_entries() failed");
+			{ errno = e; return nullptr; }
 		dir->__ent_next = 0;
 		if(!dir->__ent_limit)
 			return nullptr;
 	}
