diff --git a/options/ansi/generic/string.cpp b/options/ansi/generic/string.cpp
--- a/options/ansi/generic/string.cpp
+++ b/options/ansi/generic/string.cpp
@@ -480,7 +480,7 @@ char *strerror(int e) {
 	case EINVAL: s = "Invalid argument (EINVAL)"; break;
 	case EIO:    s = "I/O error (EIO)"; break;
 	case EISDIR: s = "Resource is directory (EISDIR)"; break;
-	case ENOENT: s = "No such file or directory (ENOENT)"; break;
+	case ENOENT: s = "No such file or directory"; break;
 	case ENOMEM: s = "Out of memory (ENOMEM)"; break;
 	case ENOTDIR: s = "Expected directory instead of file (ENOTDIR)"; break;
 	case ENOSYS: s = "Operation not implemented (ENOSYS)"; break;
