diff --git a/options/ansi/generic/stdlib.cpp b/options/ansi/generic/stdlib.cpp
--- a/options/ansi/generic/stdlib.cpp
+++ b/options/ansi/generic/stdlib.cpp
@@ -413,7 +413,18 @@ void qsort_r(void *base, size_t count, size_t size,
 	auto partition = [&] (size_t begin, size_t end) -> size_t {
 		__ensure(begin < end);
+
+		auto last = end - 1;
+		auto middle = begin + (end - begin) / 2;
+		if (compare_idx(middle, begin) < 0)
+			swap_idx(middle, begin);
+		if (compare_idx(last, begin) < 0)
+			swap_idx(last, begin);
+		if (compare_idx(last, middle) < 0)
+			swap_idx(last, middle);
+		swap_idx(middle, last);
+
 		// Invariant: every element before i is compares smaller or equal to the pivot.
-		auto pivot = end - 1;
+		auto pivot = last;
 		size_t i = begin;
 		for (size_t j = begin; j < pivot; ++j) {
 			if (compare_idx(j, pivot) <= 0) {
@@ -426,15 +437,20 @@ void qsort_r(void *base, size_t count, size_t size,
 	};
 
 	// Sort the range [begin, end).
-	// TODO: If we are doing more than C * log_2(n) iterations for some C,
-	//       fall back to a guaranteed O(log n) implementation.
-	auto quick_sort = [&] (this auto self, size_t begin, size_t end) {
+	auto quick_sort = [&] (this auto self, size_t begin, size_t end) -> void {
 		__ensure(begin <= end);
-		if (end - begin <= 1)
-			return;
-		auto pivot = partition(begin, end);
-		self(begin, pivot);
-		self(pivot + 1, end);
+		while (end - begin > 1) {
+			auto pivot = partition(begin, end);
+			auto right_begin = pivot + 1;
+
+			if (pivot - begin < end - right_begin) {
+				self(begin, pivot);
+				begin = right_begin;
+			} else {
+				self(right_begin, end);
+				end = pivot;
+			}
+		}
 	};
 
 	quick_sort(0, count);
