diff --git a/options/ansi/generic/file-io.cpp b/options/ansi/generic/file-io.cpp
--- a/options/ansi/generic/file-io.cpp
+++ b/options/ansi/generic/file-io.cpp
@@ -86,7 +86,7 @@
 }
 
 abstract_file::~abstract_file() {
-	if(__dirty_begin != __dirty_end)
+	if(__dirty_begin != __dirty_end && !(__status_bits & __MLIBC_ERROR_BIT))
 		mlibc::infoLogger() << "mlibc warning: File is not flushed before destruction"
 				<< frg::endlog;
 
@@ -481,7 +481,7 @@
 }
 
 int fd_file::close() {
-	if(__dirty_begin != __dirty_end)
+	if(__dirty_begin != __dirty_end && !(__status_bits & __MLIBC_ERROR_BIT))
 		mlibc::infoLogger() << "mlibc warning: File is not flushed before closing"
 				<< frg::endlog;
 	if(int e = mlibc::sysdep<Close>(_fd); e)
@@ -714,11 +714,17 @@
 int fclose(FILE *file_base) {
 	auto file = static_cast<mlibc::abstract_file *>(file_base);
 	int e = 0;
-	if(file->flush())
+	int flush_error = file->flush();
+	if(flush_error)
 		e = EOF;
-	if(file->close())
+	int close_error = file->close();
+	if(close_error)
 		e = EOF;
 	file->dispose();
+	if(flush_error > 0)
+		errno = flush_error;
+	else if(!flush_error && close_error > 0)
+		errno = close_error;
 	return e;
 }
 
@@ -755,8 +761,11 @@
 		return 0;
 	}
 	auto file = static_cast<mlibc::abstract_file *>(file_base);
-	if(file->flush())
+	if(int flush_error = file->flush(); flush_error) {
+		if(flush_error > 0)
+			errno = flush_error;
 		return EOF;
+	}
 	return 0;
 }
 int fflush(FILE *file_base) {
@@ -774,8 +783,11 @@
 
 	auto file = static_cast<mlibc::abstract_file *>(file_base);
 	frg::unique_lock lock(file->_lock);
-	if (file->flush())
+	if(int flush_error = file->flush(); flush_error) {
+		if(flush_error > 0)
+			errno = flush_error;
 		return EOF;
+	}
 	return 0;
 }
 
